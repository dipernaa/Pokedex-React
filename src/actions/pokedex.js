import api from './utils/api';
import { createAsyncAction } from './utils/asyncUtils';
import { GET_POKEMON } from '../constants/actionTypes';

function getPokemon() {
  return createAsyncAction(GET_POKEMON, () => api.getAsync('/api/pokedex'));
}

export default {
  getPokemon
};
