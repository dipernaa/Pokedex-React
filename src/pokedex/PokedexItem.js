import React from 'react';
import { Card, CardHeader, CardText } from 'material-ui/Card';

const PokedexItem = ({ getPokemonDetail, pokemon }) => (
  <Card
    onExpandChange={() => getPokemonDetail(pokemon.url)}
  >
    <CardHeader
      title={pokemon.name}
      actAsExpander
      showExpandableButton
    />
    <CardText expandable>
      hi
    </CardText>
  </Card>
);

export default PokedexItem;
