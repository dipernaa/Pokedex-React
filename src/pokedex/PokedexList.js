import React, { Component } from 'react';
import { browserHistory } from 'react-router';
import { List, ListItem } from 'material-ui/List';
import TextField from 'material-ui/TextField';
import { routeNames } from '../routes';

const styles = {
  container: {
    display: 'flex',
    flexDirection: 'column'
  },
  search: {
    marginLeft: 'auto',
    marginRight: 'auto'
  }
};

class PokedexList extends Component {

  constructor(props) {
    super(props);

    this.state = {
      filteredPokedex: props.pokedex
    };
  }

  filterPokedex = (value) => {
    if (value) {
      this.setState({
        filteredPokedex: this.props.pokedex.filter(item => item.name.indexOf(value) >= 0)
      });
    } else {
      this.setState({
        filteredPokedex: this.props.pokedex
      });
    }
  }

  render() {
    if (!this.state.filteredPokedex) {
      return null;
    }

    return (
      <div style={styles.container}>
        <TextField
          hintText="Search"
          onChange={(event, value) => this.filterPokedex(value)}
          style={styles.search}
        />
        <br/>

        <div>
          <List>
            {this.state.filteredPokedex.map((pokemon, index) => (
              <ListItem
                key={index}
                className="capitalize"
                onClick={() => browserHistory.push(`${routeNames.detail}?url=${pokemon.url}`)}
                primaryText={pokemon.name}
              />
            ))}
          </List>
        </div>
      </div>
    );
  }
}

export default PokedexList;
