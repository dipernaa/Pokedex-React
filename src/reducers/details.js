import { GET_POKEMON_DETAILS } from '../constants/actionTypes';

export const initialState = {
  results: null,
  isLoading: false
};

export default function(state = initialState, action) {
  console.log(state);
  console.log(action);
  switch (action.type) {
    case GET_POKEMON_DETAILS.request:
      return {
        ...state,
        isLoading: true
      };

    case GET_POKEMON_DETAILS.failure:
      return {
        ...state,
        isLoading: false
      };

    case GET_POKEMON_DETAILS.success:
      return {
        ...state,
        results: action.data,
        isLoading: false
      };

    default:
      return state;
  }
}
