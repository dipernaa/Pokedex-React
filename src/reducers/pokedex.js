import { GET_POKEMON } from '../constants/actionTypes';

const sortPokemon = (arr) => (
  arr.sort((first, second) => {
    if (first.name > second.name) {
      return 1;
    } else if (first.name < second.name) {
      return -1;
    }

    return 0;
  })
);

export const initialState = {
  count: 0,
  next: null,
  previous: null,
  results: [],
  isLoading: false
};

export default function(state = initialState, action) {
  switch (action.type) {
    case GET_POKEMON.request:
      return {
        ...state,
        isLoading: true
      };

    case GET_POKEMON.failure:
      return {
        ...state,
        isLoading: false
      };

    case GET_POKEMON.success:
      return {
        ...state,
        count: action.data.count,
        next: action.data.next,
        previous: action.data.previous,
        results: sortPokemon(action.data.results.map((item, index) => ({
          ...item,
          id: index + 1
        }))),
        isLoading: false
      };

    default:
      return state;
  }
}
