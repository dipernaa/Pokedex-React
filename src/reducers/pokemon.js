import {
  GET_POKEMON,
  GET_POKEMON_DETAILS
} from '../constants/actionTypes';

const sortPokemon = (arr) => (
  arr.sort((first, second) => {
    if (first.name > second.name) {
      return 1;
    } else if (first.name < second.name) {
      return -1;
    }

    return 0;
  })
);

export const initialState = {
  count: 0,
  next: null,
  previous: null,
  results: [],
  isLoading: false
};

export default function(state = initialState, action) {
  // console.log(state);
  // console.log(action);
  switch (action.type) {
    case GET_POKEMON.request:
    case GET_POKEMON_DETAILS.request:
      return {
        ...state,
        isLoading: false
      };

    case GET_POKEMON.failure:
    case GET_POKEMON_DETAILS.failure:
      return {
        ...state,
        isLoading: false
      };

    case GET_POKEMON.success:
      return {
        ...state,
        count: action.data.count,
        next: action.data.next,
        previous: action.data.previous,
        results: sortPokemon(action.data.results.map((item, index) => ({
          ...item,
          id: index + 1
        }))),
        isLoading: false
      };

    case GET_POKEMON_DETAILS.success:
      return {
        ...state,
        results: sortPokemon(state.results.map(item => {
          if (item.id === action.data.id) {
            return {
              ...item,
              details: action.data
            };
          }

          return {
            ...item
          };
        })),
        isLoading: false
      };

    default:
      return state;
  }
}
