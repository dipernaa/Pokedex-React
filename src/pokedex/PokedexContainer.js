import React, { Component } from 'react';
import { connect } from 'react-redux';
import muiThemeable from 'material-ui/styles/muiThemeable';
import PokedexItem from './PokedexItem';
import Loading from '../shared/Loading';
import { getPokemon, getPokemonDetails } from '../actions';

export class PokedexContainer extends Component {

  componentWillMount() {
    this.props.getPokemon();
  }

  render() {
    const { getPokemonDetail, isLoading, pokemonData } = this.props;

    if (isLoading) {
      return <Loading />;
    }

    return (
      <div>
        {pokemonData.map((pokemon, index) => (
          <PokedexItem
            key={index}
            isLoading={isLoading}
            getPokemonDetail={getPokemonDetail}
            pokemon={pokemon}
          />
        ))}
      </div>
    );
  }
}

export const mapStateToProps = (state) => {
  const { pokemon } = state;

  return {
    isLoading: pokemon.isLoading,
    pokemonData: pokemon.results
  };
};

export const mapDispatchToProps = (dispatch) => ({
  getPokemon: () => dispatch(getPokemon()),
  getPokemonDetail: (url) => dispatch(getPokemonDetails(url))
});

export default muiThemeable()(connect(mapStateToProps, mapDispatchToProps)(PokedexContainer));
