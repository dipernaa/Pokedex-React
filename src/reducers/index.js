import { combineReducers } from 'redux';
import details from './details';
import pokedex from './pokedex';

export default combineReducers({
  details,
  pokedex
});
