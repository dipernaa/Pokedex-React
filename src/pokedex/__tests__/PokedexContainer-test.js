import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import mockStore from 'redux-mock-store';
import { PokedexContainer, mapStateToProps, mapDispatchToProps } from '../PokedexContainer';
import { initialState } from '../../reducers/pokedex';

const mockReduxStore = mockStore({});
const TEST_PROPS = {
  getPokemon: jest.fn()
};

beforeEach(() => {
  mockReduxStore.clearActions();
});

it('renders a PokedexList when not loading', () => {
  const props = {
    ...TEST_PROPS,
    isLoading: false
  };

  const wrapper = shallow(
    <PokedexContainer {...props}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders a Loading bar when loading', () => {
  const props = {
    ...TEST_PROPS,
    isLoading: true
  };

  const wrapper = shallow(
    <PokedexContainer {...props}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('correctly maps state to props', () => {
  const state = {
    pokedex: initialState
  };

  const props = mapStateToProps(state);
  expect(props).toMatchSnapshot();
});

it('correctly maps dispatch to props', () => {
  const props = mapDispatchToProps(mockReduxStore.dispatch);
  expect(props).toMatchSnapshot();

  const testResponse = { leap: 'rules' };

  fetch.mockResponseSuccess(testResponse);
  return props.getPokemon()
    .then(() => {
      expect(mockReduxStore.getActions()).toMatchSnapshot();
    });
});
