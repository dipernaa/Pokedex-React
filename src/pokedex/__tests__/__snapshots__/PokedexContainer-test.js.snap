// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly maps dispatch to props 1`] = `
Object {
  "getPokemon": [Function],
}
`;

exports[`correctly maps dispatch to props 2`] = `
Array [
  Object {
    "type": "GET_POKEMON_REQUEST",
  },
  Object {
    "leap": "rules",
    "type": "GET_POKEMON_SUCCESS",
  },
]
`;

exports[`correctly maps state to props 1`] = `
Object {
  "isLoading": false,
  "pokedexData": Array [],
}
`;

exports[`renders a Loading bar when loading 1`] = `<Loading />`;

exports[`renders a PokedexList when not loading 1`] = `<PokedexList />`;
